src/nuke/GradeBeautyLayerSet.cpp

```diff
@@ -60,59 +61,51 @@ class GradeBeautyLayerSet : public PixelIop {
     bool clampBlack{true};
     bool clampWhite{false};
     int m_operation{operationModes::ADD};
-    LayerSetKnobWrapper layerSetKnob;
+    LayerSetKnobData m_lsKnobData;
     ChannelSet m_targetLayer{Mask_RGB};
     // intermediate grade algorithm storage
     float A[3]{0.0f, 0.0f, 0.0f};
     float B[3]{0.0f, 0.0f, 0.0f};
     float G[3]{0.0f, 0.0f, 0.0f};
 
 public:
-
-    bool pass_transform() const {
-        return true;
-    }
-    void append(Hash&);
+    void knobs(Knob_Callback);
     void _validate(bool for_real);
+    bool pass_transform() const {return true;}
     void in_channels(int, ChannelSet& channels) const;
     void pixel_engine(const Row&, int, int, int, ChannelMask, Row&);
-    void knobs(Knob_Callback);
     int knob_changed(Knob*);
+    const char* Class() const {return description.name;}
+    const char* node_help() const {return HELP;}
+    static const Iop::Description description;
+
     float validateGammaValue(const float&);
     // This function calculates and stores the grade algorithm's intermediate calculations
     void precomputeValues();
-
-    const char* Class() const {
-        return CLASS;
-    }
-
-    const char* node_help() const {
-        return HELP;
-    }
-    static const Iop::Description description;
+    // channel set that contains all channels that are modified by the node
+    ChannelSet activeChannelSet() const {return ChannelSet(m_targetLayer + m_lsKnobData.m_selectedChannels);}
 
     GradeBeautyLayerSet(Node* node) : PixelIop(node) { //printf("created GradeBeautyLayerSet %p\n", (void*) this);
         precomputeValues();
     }
-
-    ~GradeBeautyLayerSet() {
-    }
+    ~GradeBeautyLayerSet();
 };
 
-void GradeBeautyLayerSet::append(Hash& hash) {
-    hash.append(m_operation);
-    hash.append(layerSetKnob.configuredLayerSetName());
-}
-
-static Iop* build(Node* node) {
-    return (new NukeWrapper(new GradeBeautyLayerSet(node)))->noChannels()->mixLuminance();
 
+static Op* build(Node* node) {
+    return (new NukeWrapper(new GradeBeautyLayerSet(node)))->noChannels();
```

Comment:
When are the new NukeWrapper and GradeBeautyLayerSet destroyed ?

Reply:
Good question, from my research, and testing this PR, it really looks like constructor/destructors are handled by Nuke with the Op class.

This PR is all about trying to simplify the code, because Nuke does (probably smart) memory management that, in practice, is better not to interfere with.

any advice welcome as always, thank you @charlesfleche :)

from [](https://learn.foundry.com/nuke/developers/120/ndkreference/Plugins/classDD_1_1Image_1_1Op.html)

> Detailed Description 
> 
> Base class of all the things that can be created by Nuke nodes.
> 
> Nuke will create these things, hook their inputs to other ones, manage animation and store animated values into these by calling the knobs() function, **and delete these when the node is destroyed.**
> 
> **Nuke will also invisibly manage multiple copies hidden inside a single node**, where each copy is at a different time. This allows a later operator to ask for several time samples and merge them together in order to get motion blur.